keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,1235 @@
+- /pkg/kubelet/cm/cpumanager/policy_options.go
+- /pkg/kubelet/cm/cpumanager/topology/topology.go
+
+Existing topoology test cases will be modified to include uncorecache topology. All modified and added functions will have new test cases.
+
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+- `/pkg/kubelet/cm/cpumanager/cpu_assignment.go`: `2024-08-26` - `93.5%`
+- `/pkg/kubelet/cm/cpumanager/cpu_manager.go`: `2024-08-26` - `74.5%`
+- `/pkg/kubelet/cm/cpumanager/policy_static.go`: `2024-08-26` - `89.5%`
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+Will follow existing integration testing required in PR check.
+
+##### e2e tests
+
+- e2e testing should cover the deployment of a pod that is following LLC alignment which can be verified in the CPU Manager State file
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- For e2e testing, guaranteed pods will be deployed with various CPU size requirements on our own baremetal instances across different vendor architectures and confirming the CPU assignments to uncore cache core groupings. This feature is intended for baremetal only and not cloud instances.
+- Update CI to test GCP instances of different architectures utilizing uncore cache alignment feature.
```

Comment:
I'm not saying that shouldn't be a test but that doesn't bring a lot of value, that's more like a regression test

Reply:
Need clarification on this.
For e2e testing, I believe that we are being asked to create specific tests for when the feature is enabled since it is not intended to be on by default. Existing e2e test cases exist just to ensure it builds successfully without any feature gates enabled.
For feature specific e2e testing, we would need to create a general function that creates the topology of the node with CacheIDs, deploy guaranteed pods, and confirm CPU assignments to proper uncore cache.

Reply:
Let's take this example: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/topology_manager_test.go#L893

The best we can do is write the e2e tests as if there is a CPU with split L3, but add checks at the beginning of the tests to actually verify the hardware capabilities. If the hardware don't meet the requirements, skip.
This way we can run on baremental, and enable the tests in CI, which runs on cloud instances, and the tests will do the right thing.

Unless there's no other option left, I'm against faking the topology or any other cluster property. Doing so it kinda defies the point of e2e testing.

All that said, e2e tests are a beta-level requirement but "just" a nice to have for alpha-level features. Of course, the sooner we have the e2e tests the better, and we usually _try_ hard to have them for alpha.